### DISCUSSION TITLE
invisible text layer

### DISCUSSION THREAD

**Discussion by lbr991** [pymupdf/PyMuPDF#2464] [Q&A]
I'm trying to take a non-searchable pdf and convert it to searchable pdf by superimposing an invisible text layer. 

        
            #Parse blocks received from textract response
            blocks = response['Blocks']
            
            for block in blocks:
                if block['BlockType'] == 'WORD':
                    page_number = block['Page'] - 1  # pages in PyMuPDF start from 0
                    pdf_page = doc.load_page(page_number)
            
                    bbox = block['Geometry']['BoundingBox']
                    bbox_mupdf = fitz.Rect(
                        bbox['Left'] * pdf_page.rect.width,
                        bbox['Top'] * pdf_page.rect.height,
                        (bbox['Left'] + bbox['Width']) * pdf_page.rect.width,
                        (bbox['Top'] + bbox['Height']) * pdf_page.rect.height
                    )
                    
                    pdf_page.insert_textbox(
                        rect=bbox_mupdf,
                        buffer=block['Text'],
                        color=None,  # Invisible color
                        overlay=True  # Overlay the text on top of existing content
                    )

I want the invisible bboxes to align exactly with the original bboxes (extracted by Textract), but with this code the resulting bboxes are off (smaller). 
Is insert_textbox the right way to do this? 
I don't want to specify a font and font_size because then the bboxes wouldn't align perfectly.

**Comment by JorjMcKie** [comment]
There are a few options to improve things:
* Textract coordinates are relative to a page rectangle `(0, 0, 1, 1)`. So you have to transform every Textract bbox to the bbox of the page.
* This works best by computing a matrix that does this transformation: `matrix = fitz.Rect(0, 0, 1, 1).torect(page.rect)`.
* For every bbox in Textract you need do this to compute the bbox on the page: `bbox_mupdf = fitz.Rect(
bbox['Left'], bbox['Top'], bbox['Left'] + bbox['Width'], bbox['Top'] + bbox['Height']) * matrix
* Textract also does not deliver fonts, so you must decide about one. Suggestion: Helvetica, i.e. `font = fitz.Font("helv")`.
* Before you can insert text, you must compute the font size, whereas you only have `bbox.width`. So do `tl = font.text_length(text, fontsize=1)` with the text of e.g. the WORD. Then compute font size = bbox.width / tl.
* the insertion **_point_** is the next issue: if text contains no descender characters below the baseline, choose `bbox.bl` as insertion point: `page.insert_text(bbox.bl, text, fontsizefontsize), render_mode=3)`.
* If a character extending below baseline exists (y, g, ...), then the insertion point must be a little higher: `bbox.bl + (0, font.descender*fontsize)` (font.descender is negative).
* In any case use ender_mode=3, which creates invisible text - no need to fiddle around with text colors.
* In any case do **not use insert_textbox()**, but insert_text() as described above.

As an aside: this is a typcal "Discussions" item, so I will transfer it.

**Reply by lbr991** [reply]
Thanks for the prompt response!
Is there a way to make the invisible bbox stretch to the original bbox size (so that they perfectly align)? By using Helvetica the resulting bbox might similar but won't be the exact bbox, unless I misunderstand something.

**Reply by JorjMcKie** [reply]
Of course, it may happen - depending on the original (forever unknown) font - stretching the text to at least exhaust the bbox width may lead to an awkward height: exceeding bbox.height.
There can be nothinh done about this effect - escept trying out a few different fonts, Times-Roman etc.
Example calculation:
```python 
helv = fitz.Font("helv")  # font example
tiro = fitz.Font("tiro")  # font example
bbox= fitz.Rect(100,100,200,20)  # bbox of the text
text = "Jorj X. Mckie"  # the text
tl_helv = font.text_length(text,fontsize=1)
tl_tiro = tiro.text_length(text,fontsize=1)

bbox.width/tl_helv  # fontsize for helv
17.47335316704016
bbox.width/tl_tiro  # fontsize for Times Roman
18.092997976990105
(font.ascender-font.descender)*17.473353  # line height for Helvetica
24.008387767708985
(tiro.ascender-tiro.descender)*18.093  # line height for Times Roman
24.136061305493115
```
The insertion point for Helvetica would have to be (note the descending character "j") `bbox.bl + (0, delta)`, where `delta = helv.descender / (helv.ascender - helv.descender) *17.473353 = -3.8`.
